### PR TITLE
PeerInfo for p2p

### PR DESCRIPTION
**Pull request by leviathanbeak**
- added PeerInfoBehaviour
- moved some stuff from DiscoveryBehaviour to PeerInfoBehaviour
- improved and refactored the tests

### CODE REVIEW

**Comment by Voxelot** [comment]
@leviathanbeak looks like this needs a rebase with master branch.

File: fuel-p2p/src/behavior.rs [hunk 1]
```diff
@@ -12,6 +15,9 @@ use std::{
     collections::VecDeque,
     task::{Context, Poll},
 };
+use tracing::debug;
+
+const MAX_IDENTIFY_ADDRESSES: usize = 10;
```
Comment (Voxelot) [inline fuel-p2p/src/behavior.rs:20]: could you add a short comment explaining what this constant is used for?

File: fuel-p2p/src/peer_info.rs [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+                }
+                Poll::Ready(NetworkBehaviourAction::GenerateEvent(ping_event)) => {
+                    if let PingEvent {
+                        peer,
+                        result: Ok(PingSuccess::Ping { rtt }),
+                    } = ping_event
+                    {
+                        self.insert_latest_ping(&peer, rtt)
+                    }
+                }
```
Comment (Voxelot) [inline fuel-p2p/src/peer_info.rs:228]: the if-let can be combined with the match statement, ie

```suggestion
                Poll::Ready(NetworkBehaviourAction::GenerateEvent(PingEvent {
                        peer,
                        result: Ok(PingSuccess::Ping { rtt }),
                })) => {
                    self.insert_latest_ping(&peer, rtt)
                }
```

File: fuel-p2p/src/peer_info.rs [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+    /// Insert latest ping to a connected Node
+    fn insert_latest_ping(&mut self, peer_id: &PeerId, duration: Duration) {
+        match self.peers.get_mut(peer_id) {
+            Some(peer_info) => {
+                peer_info.latest_ping = Some(duration);
+            }
+            _ => {
+                debug!(target: "fuel-libp2p", "Peer with PeerId: {:?} is not among the connected peers", peer_id)
+            }
+        }
```
Comment (Voxelot) [inline fuel-p2p/src/peer_info.rs:120]: if let might be cleaner than match here

```suggestion
        if let Some(peer_info) = self.peers.get_mut(peer_id) {
            peer_info.latest_ping = Some(duration);
         } else {
             debug!(target: "fuel-libp2p", "Peer with PeerId: {:?} is not among the connected peers", peer_id)
        }
```

File: fuel-p2p/src/service.rs [hunk 4]
```diff
@@ -6,7 +6,7 @@ use futures::prelude::*;
 use libp2p::{identity::Keypair, multiaddr::Protocol, swarm::SwarmEvent, Multiaddr, PeerId, Swarm};
 use std::error::Error;
 
-/// Listens to evets on the p2p network
+/// Listens to the evets on the p2p network
```
Comment (Voxelot) [inline fuel-p2p/src/service.rs:9]:
```suggestion
/// Listens to the events on the p2p network
```

**Comment by Voxelot** [comment]
It seems like these tests are mainly related to discovery (verifying mdns behavior), but there's no test to verify the peer info. E.g. last_ping gets updates, client versions are correct, etc.

**Comment by leviathanbeak** [comment]
> It seems like these tests are mainly related to discovery (verifying mdns behavior), but there's no test to verify the peer info. E.g. latest_ping gets updates, client versions are correct, etc.

The main use case of PeerInfo at the moment is to identify peer's addresses once a peer has been discovered,
that main use case is tested in the test `nodes_connected_via_identify()`

But I will try and add some other miscellaneous tests

**Review by Voxelot — APPROVED** [review]
The new test is a little slow (15-17s) but seems stable after many runs.

**Pull request merged**